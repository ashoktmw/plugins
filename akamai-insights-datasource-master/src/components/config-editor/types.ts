export enum Secret {
  ClientSecret = 'clientSecret',
  Host = 'host',
  AccessToken = 'accessToken',
  ClientToken = 'clientToken'
}

export const secretsNames = [ Secret.ClientSecret, Secret.Host, Secret.AccessToken, Secret.ClientToken ];

export const labelWidth = 20;
export const inputWidth = 60;
