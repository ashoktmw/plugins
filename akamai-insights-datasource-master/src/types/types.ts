import { DataSourceJsonData, DataSourcePluginOptionsEditorProps } from '@grafana/data';
import { DataQuery } from '@grafana/schema';

export interface MyQuery extends DataQuery {
  reportLink?: string;
  dimensions?: string[];
  metrics?: string[];
  filters?: FilterQuery[];
  sortBys?: SortByQuery[];
}

export interface FilterQuery {
  metricName?: string;
  dimensionName?: string;
  operator?: string;
  expressions?: ExpressionType[];
  expression?: ExpressionType;
}

export interface SortByQuery {
  name?: string;
  sortOrder?: string;
}

export type ExpressionType = string | number;

export interface MyDataSourceOptions extends DataSourceJsonData {
  clientSecret?: string;
  host?: string;
  accessToken?: string;
  clientToken?: string;
}

export interface DataSourceProps extends DataSourcePluginOptionsEditorProps<MyDataSourceOptions> {}

export enum TestDataSourceResponseStatus {
  Success = 'success',
  Error = 'error'
}

export interface NameAware {
  name: string
}

export interface Enum {
  [id: string]: string;
}

export const API_URL = '/reporting-api/v2/reports/';

export const VARIABLE_QUERY = 'variable-query';
