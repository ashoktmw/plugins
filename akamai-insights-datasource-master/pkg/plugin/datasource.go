package plugin

import (
	"bytes"
	"context"
	"encoding/json"
	"github.com/grafana/grafana-plugin-sdk-go/backend"
	"github.com/grafana/grafana-plugin-sdk-go/backend/datasource"
	"github.com/grafana/grafana-plugin-sdk-go/backend/instancemgmt"
	"net/http"
	"net/url"
)

type DataRequest struct {
	Body RequestBody `json:"body"`
	From string      `json:"from"`
	To   string      `json:"to"`
}

type RequestBody struct {
	Dimensions []string                 `json:"dimensions"`
	Metrics    []string                 `json:"metrics"`
	Filters    []map[string]interface{} `json:"filters"`
	SortBys    []map[string]interface{} `json:"sortBys"`
}

type DataSourceSettings struct {
	ClientSecret string `json:"clientSecret"`
	Host         string `json:"host"`
	AccessToken  string `json:"accessToken"`
	ClientToken  string `json:"clientToken"`
}

func newDataSourceInstance(ctx context.Context, setting backend.DataSourceInstanceSettings) (instancemgmt.Instance, error) {
	var dss DataSourceSettings
	err := json.Unmarshal(setting.JSONData, &dss)

	if err != nil {
		return 1, err
	}

	return &instanceSettings{
		httpClient: &http.Client{},
	}, nil
}

type instanceSettings struct {
	httpClient *http.Client
}

func NewDatasource() datasource.ServeOpts {
	// Creates a instance manager for the plugin. The function passed
	// into `NewInstanceManger` is called when the instance is created
	// for the first time or when datasource configuration changes.
	im := datasource.NewInstanceManager(newDataSourceInstance)

	ds := &AkamaiEdgeDnsDatasource{
		im: im,
	}

	return datasource.ServeOpts{
		CallResourceHandler: ds,
	}
}

type AkamaiEdgeDnsDatasource struct {
	im instancemgmt.InstanceManager
}

func (td *AkamaiEdgeDnsDatasource) CallResource(ctx context.Context, req *backend.CallResourceRequest, sender backend.CallResourceResponseSender) error {
	var dataSourceSettings DataSourceSettings

	if err := json.Unmarshal(req.PluginContext.DataSourceInstanceSettings.JSONData, &dataSourceSettings); err != nil {
		return err
	}

	switch req.Path {
	case "reports":
		query, err := reportApiQuery(dataSourceSettings)

		if err != nil {
			return err
		}

		body, err := json.Marshal(query)

		if err != nil {
			return err
		}

		return sender.Send(&backend.CallResourceResponse{
			Status: http.StatusOK,
			Body:   body,
		})
	case "discovery":
		parse, err := url.Parse(req.URL)
		if err != nil {
			return err
		}
		var targetUrl = parse.Query().Get("targetUrl")

		query, err := discoveryApiQuery(dataSourceSettings, targetUrl)

		if err != nil {
			return err
		}

		body, err := json.Marshal(query)

		if err != nil {
			return err
		}

		return sender.Send(&backend.CallResourceResponse{
			Status: http.StatusOK,
			Body:   body,
		})
	case "data":
		parse, err := url.Parse(req.URL)
		if err != nil {
			return err
		}
		var targetUrl = parse.Query().Get("targetUrl")

		var requestData DataRequest
		err = json.Unmarshal(req.Body, &requestData)

		if err != nil {
			return err
		}

		marshal, err := json.Marshal(requestData.Body)

		if err != nil {
			return err
		}

		b := bytes.NewBuffer(marshal)
		query, errorResponse := dataQuery(dataSourceSettings, targetUrl, b, requestData.From, requestData.To)

		if errorResponse != nil {
			var readableError, _ = json.Marshal(errorResponse)
			return sender.Send(&backend.CallResourceResponse{
				Status: errorResponse.Status,
				Body:   readableError,
			})
		}

		body, err := json.Marshal(query)

		if err != nil {
			return err
		}

		return sender.Send(&backend.CallResourceResponse{
			Status: http.StatusOK,
			Body:   body,
		})
	default:
		return sender.Send(&backend.CallResourceResponse{
			Status: http.StatusNotFound,
			Body:   []byte("Unexpected resource URI: " + req.Path),
		})
	}
}
