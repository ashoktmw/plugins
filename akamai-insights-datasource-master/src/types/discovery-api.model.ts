import { DataQuery } from '@grafana/schema';

export interface DiscoveryApiModel extends DataQuery {
  metrics: Metric[];
  dimensions: Dimension[];
  defaults: Defaults;
}

export interface ReportsApiModel {
  reports: Report[]
}

export interface Report {
  reportName: string;
  reportLink: string;
}

export interface Metric {
  name: string;
  type: DataType;
}

export interface Dimension {
  name: string;
  type: DataType;
  filterable: boolean;
  filterType: string;
  authorizable: boolean;
  filterEnumValues?: string[];
}

export interface Defaults {
  defaultTimeRange?: DefaultTimeRange;
  defaultMetrics?: string[];
  defaultDimensions?: string[];
  defaultSortBys?: DefaultSortBy[];
}

export interface DefaultTimeRange {
  start: string;
  end: string;
}

export interface DefaultSortBy {
  name: string;
  sortOrder: 'ASCENDING' | 'DESCENDING';
}

export const initialModel: DiscoveryApiModel = {
  metrics: [],
  dimensions: [],
  defaults: {},
  refId: ''
};

export enum DataType {
  STRING = 'STRING',
  LONG = 'LONG',
  DOUBLE = 'DOUBLE',
  TIMESTAMP_SEC = 'TIMESTAMP_SEC',
  DATE_ISO8601 = 'DATE_ISO8601',
  TIMESTAMP_MS = 'TIMESTAMP_MS'
}

export const TimeDimensionsTypes = [ DataType.TIMESTAMP_SEC, DataType.DATE_ISO8601, DataType.TIMESTAMP_MS ];
