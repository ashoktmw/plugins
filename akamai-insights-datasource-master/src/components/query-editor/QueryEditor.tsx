import { QueryEditorProps } from '@grafana/data';
import { Cascader, CascaderOption, Icon } from '@grafana/ui';
import { isEmpty } from 'lodash';
import React, { useLayoutEffect, useState } from 'react';

import { DataSourceForm } from './DataSourceForm/DataSourceForm';
import { DatasourceService } from '../../services/datasource.service';
import { DatasourcesCascaderService } from '../../services/datasources-cascader.service';
import { initialModel } from '../../types/discovery-api.model';
import { MyDataSourceOptions, MyQuery } from '../../types/types';

type Props = QueryEditorProps<DatasourceService, MyQuery, MyDataSourceOptions>;

export function QueryEditor({ query, onChange, onRunQuery, datasource }: Props) {
  const cascaderSeparator = ' > ';
  const [ model, setState ] = useState(initialModel);
  const [ dataSource, setDataSource ] = useState<string>('');
  const [ isDataSourceLoading, setIsDataSourceLoading ] = useState(false);
  const [ isReportListLoading, setIsReportListLoading ] = useState(true);
  const [ cascaderOptions, setCascaderOptions ] = useState([] as CascaderOption[]);

  const onDataSourceOptionChange = (value: string) => {
    query.reportLink = value;
    setDataSource(value || '');

    if (value) {
      setIsDataSourceLoading(true);
      DatasourceService.discoveryApi(datasource.id, value).subscribe({
        next: data => setState(data),
        complete: () => setIsDataSourceLoading(false)
      });
    }
  };

  useLayoutEffect(() => {
    const subscription = DatasourceService.reportsApi(datasource.id).subscribe({
      next: data => {
        setCascaderOptions(DatasourcesCascaderService.getCascaderOptions(data));
        onDataSourceOptionChange(query.reportLink || '');
        setIsReportListLoading(false);
      },
      error: () => {
        setCascaderOptions([]);
        setIsReportListLoading(false);
      }
    });

    return () => subscription.unsubscribe();
    // eslint-disable-next-line react-hooks/exhaustive-deps
  }, [ datasource.id ]);

  return (
    <>
      {(() => !isReportListLoading ?
        <Cascader
          initialValue={dataSource ? DatasourcesCascaderService.getDataSourceParts(dataSource).join(cascaderSeparator) : ''}
          placeholder="Select report data source"
          allowCustomValue={true}
          options={cascaderOptions}
          changeOnSelect={false}
          onSelect={onDataSourceOptionChange}
          displayAllSelectedLevels={true}
          separator={cascaderSeparator}
        />
        : <><Icon name="fa fa-spinner"/> Fetching...</>
      )()}
      <hr/>
      {(() => !isDataSourceLoading && !isEmpty(model.metrics) && !isEmpty(model.dimensions) ?
        <DataSourceForm
          model={model}
          query={query}
          onChange={onChange}
          onRunQuery={onRunQuery}
          datasource={datasource}
        /> :
        isDataSourceLoading ? <><Icon name="fa fa-spinner"/> Fetching...</> : ''
      )()}
    </>
  );
}
