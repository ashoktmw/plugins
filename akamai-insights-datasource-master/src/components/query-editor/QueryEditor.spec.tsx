import React from 'react';
import renderer from 'react-test-renderer';
import { of } from 'rxjs';

import { QueryEditor } from './QueryEditor';
import { DatasourceService } from '../../services/datasource.service';
import { mockReportsApi } from '../../test/mocks/mock-reports-api';
import { MyQuery } from '../../types/types';

const datasource: any = {
  id: 1
};

describe('given QueryEditor', () => {
  test('should render new empty editor', () => {
    const query = {
      dimensions: [],
      metrics: [],
      reportLink: ''
    } as any as MyQuery;

    jest.spyOn(DatasourceService, 'reportsApi').mockReturnValue(of(mockReportsApi));
    const onChange = jest.fn();
    const onRunQuery = jest.fn();

    const tree = renderer.create(
      <QueryEditor query={query} datasource={datasource} onChange={onChange} onRunQuery={onRunQuery}/>
    ).toJSON();

    expect(tree).toMatchSnapshot();
  });

  test('should render editor with initial value', () => {
    const query = {
      dimensions: [],
      metrics: [],
      reportLink: '/reporting-api/v2/reports/delivery/traffic/emissions'
    } as any as MyQuery;

    jest.spyOn(DatasourceService, 'reportsApi').mockReturnValue(of(mockReportsApi));
    const onChange = jest.fn();
    const onRunQuery = jest.fn();

    const tree = renderer.create(
      <QueryEditor query={query} datasource={datasource} onChange={onChange} onRunQuery={onRunQuery}/>
    ).toJSON();

    expect(tree).toMatchSnapshot();
  });
});
