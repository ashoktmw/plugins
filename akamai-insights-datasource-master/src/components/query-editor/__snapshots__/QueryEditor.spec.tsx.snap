// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`given QueryEditor should render editor with initial value 1`] = `
[
  <i
    className="fa fa-spinner fa-spin"
  />,
  " Fetching...",
  <hr />,
  <i
    className="fa fa-spinner fa-spin"
  />,
  " Fetching...",
]
`;

exports[`given QueryEditor should render new empty editor 1`] = `
[
  <div>
    <div
      className="css-4339yg"
      onClick={[Function]}
    >
      <div
        className="css-xcstkt-input-wrapper"
        data-testid="input-wrapper"
      >
        <div
          className="css-1w5c5dq-input-inputWrapper"
        >
          <input
            className="css-1n6y0bv-input-input"
            onBlur={[Function]}
            onChange={[Function]}
            onKeyDown={[Function]}
            placeholder="Select report data source"
            style={
              {
                "paddingLeft": undefined,
                "paddingRight": 12,
              }
            }
            value=""
          />
          <div
            className="css-7y3u6k-input-suffix"
          >
            <div
              className="css-wf08df-Icon"
            >
              <svg
                data-testid="angle-down"
                viewBox="0 0 24 24"
                xmlns="http://www.w3.org/2000/svg"
              />
            </div>
          </div>
        </div>
      </div>
    </div>
  </div>,
  <hr />,
]
`;
